restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java

```diff
@@ -212,6 +213,7 @@ public QueryExecutor(DataStore dataStore) {
                 });
 
     return RxUtils.singleFromFuture(supplier)
+        .observeOn(Schedulers.io())
```

Comment:
As discussed, we should certainly clean up thread management in Docs API, but this small change may not be sufficient.

I suppose C* OSS should generally work fine regardless of how we switch threads. The DSE use cases are probably going to be the most tricky, though.

Points to consider:
* DSE core (aka TPC) threads should never be blocked by I/O or `.sleep()` or `CompletableFuture.join()`, etc.
* Blocking I/O should be performed on another thread pool. This probably means we have to write HTTP responses on another thread pool or at least verify whether the response can be written without block from a DSE core thread. I'm not sure how exactly this works with async servlets. Please double check.
* `DataStore` queries will run callbacks on DSE core threads initially. If we have to switch to another thread pool, we should be careful with performance implications since DSE core thread generally occupy almost all CPUs under load (CPU count - 1, IIRC).
* If we use another thread pool it might be worth adding controls for min/max pool size parameters (to avoid creating too many threads under load)

Reply:
Nice results @ivansenic !

I believe it would also be valuable to review perf. under high load, though. I'm not sure what the overall CPU utilization was on the test system(s) above, but in general DSE-specific behaviours are likely to surface under high load on the coordinator, plus multiple storage nodes.

Also, single (or few) rows reads are likely to be fulfilled with a single inter-node message, which will probably not trigger any on the deep DSE-specific behaviours. It might be worth testing with complex `WHERE` conditions that require range scans or full scans to have a more complete picture.

Reply:
I'll create tests and add results for using the `where`. I'll target even larger documents and match many of them (max `page-size`).

Reply:
Run a quick search scenario as well:

**Setup:**
* Local test
* Stargate single node, with DSE single node on same machine
* Measuring search of 10k documents, each having 23 fields (rows in the database)
* First filters grabs app. 5% of the document, second grabs app. 10% of the candidates
* Results (not that absolute numbers don't matter here, this was run with 120 threads all on the single node)
   * master:  p90 ~6.95s, p99.9 ~11.89s
   * 1151:  p90 ~6.95ms, p99.9 ~13.68ms

*master*
![image](https://user-images.githubusercontent.com/10600041/129546753-d46d1506-b2b2-4b24-b33c-5d753efb75c4.png)

*#1151*
![image](https://user-images.githubusercontent.com/10600041/129546783-8eed82e9-bfe4-40d4-b07a-9538e20ac564.png)

Reply:
One more point, sorry for keeping this thread long :)

Have you considered switching to the I/O scheduler at the end of the chain? I suppose it is only really required when we _write_ the HTTP response. Filtering and posting nested queries can probably be done on the thread that processes previous query results (i.e. DSE core threads). I suppose it might help to reduce the number of context switches in highly filtered cases. WDYT?

Reply:
@dimas-b The HTTP write was not my only concern, it's also to protect that we don't do anything in the core threads that can lead to the problems we had seen with the delete leafs.. You are most likely write, perf might be best if we do the switch prior to the HTTP write, but I am also looking for that safety.